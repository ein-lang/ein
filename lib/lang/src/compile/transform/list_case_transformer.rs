use super::super::{
    error::CompileError, list_type_configuration::ListTypeConfiguration,
    name_generator::NameGenerator, reference_type_resolver::ReferenceTypeResolver,
};
use crate::{ast::*, types};
use std::sync::Arc;

pub struct ListCaseTransformer {
    first_rest_name_generator: NameGenerator,
    element_name_generator: NameGenerator,
    reference_type_resolver: Arc<ReferenceTypeResolver>,
    configuration: Arc<ListTypeConfiguration>,
}

impl ListCaseTransformer {
    pub fn new(
        reference_type_resolver: Arc<ReferenceTypeResolver>,
        configuration: Arc<ListTypeConfiguration>,
    ) -> Arc<Self> {
        Self {
            first_rest_name_generator: NameGenerator::new("$lc$firstRest"),
            element_name_generator: NameGenerator::new("$lc$firstRest"),
            reference_type_resolver,
            configuration,
        }
        .into()
    }

    pub fn transform(&self, case: &ListCase) -> Result<Expression, CompileError> {
        let source_information = case.source_information();
        let first_rest_type = types::Reference::new(
            &self.configuration.first_rest_type_name,
            source_information.clone(),
        );
        let none_type = types::None::new(source_information.clone());
        let element_type = self
            .reference_type_resolver
            .resolve_to_list(case.type_())?
            .unwrap()
            .element()
            .clone();

        let first_rest_name = self.first_rest_name_generator.generate();
        let element_name = self.element_name_generator.generate();

        Ok(Case::with_type(
            types::Union::new(
                vec![first_rest_type.clone().into(), none_type.clone().into()],
                source_information.clone(),
            ),
            &first_rest_name,
            Application::new(
                Variable::new(
                    &self.configuration.deconstruct_function_name,
                    source_information.clone(),
                ),
                case.argument().clone(),
                source_information.clone(),
            ),
            vec![
                Alternative::new(none_type, case.empty_alternative().clone()),
                Alternative::new(
                    first_rest_type,
                    Let::new(
                        vec![
                            VariableDefinition::new(
                                case.first_name(),
                                Case::with_type(
                                    types::Any::new(source_information.clone()),
                                    &element_name,
                                    Application::new(
                                        Variable::new(
                                            &self.configuration.first_function_name,
                                            source_information.clone(),
                                        ),
                                        Variable::new(&first_rest_name, source_information.clone()),
                                        source_information.clone(),
                                    ),
                                    vec![Alternative::new(
                                        element_type.clone(),
                                        Variable::new(&element_name, source_information.clone()),
                                    )],
                                    source_information.clone(),
                                ),
                                element_type,
                                source_information.clone(),
                            )
                            .into(),
                            VariableDefinition::new(
                                case.rest_name(),
                                Application::new(
                                    Variable::new(
                                        &self.configuration.rest_function_name,
                                        source_information.clone(),
                                    ),
                                    Variable::new(&first_rest_name, source_information.clone()),
                                    source_information.clone(),
                                ),
                                case.type_().clone(),
                                source_information.clone(),
                            )
                            .into(),
                        ],
                        case.non_empty_alternative().clone(),
                        source_information.clone(),
                    ),
                ),
            ],
            source_information.clone(),
        )
        .into())
    }
}

#[cfg(test)]
mod tests {
    use super::{super::super::list_type_configuration::LIST_TYPE_CONFIGURATION, *};
    use crate::debug::*;

    fn create_list_case_transformer() -> Arc<ListCaseTransformer> {
        ListCaseTransformer::new(
            ReferenceTypeResolver::new(&Module::dummy()),
            LIST_TYPE_CONFIGURATION.clone(),
        )
    }

    #[test]
    fn transform() {
        insta::assert_debug_snapshot!(create_list_case_transformer().transform(&ListCase::new(
            Variable::new("xs", SourceInformation::dummy()),
            types::List::new(
                types::Number::new(SourceInformation::dummy()),
                SourceInformation::dummy(),
            ),
            "y",
            "ys",
            None::new(SourceInformation::dummy()),
            None::new(SourceInformation::dummy()),
            SourceInformation::dummy(),
        )));
    }
}
