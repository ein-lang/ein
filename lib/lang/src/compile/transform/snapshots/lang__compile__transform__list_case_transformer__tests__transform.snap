---
source: lib/lang/src/compile/transform/list_case_transformer.rs
expression: "create_list_case_transformer().transform(&ListCase::new(Variable::new(\"xs\",\n                                                                      SourceInformation::dummy()),\n                                                        types::List::new(types::Number::new(SourceInformation::dummy()),\n                                                                         SourceInformation::dummy()),\n                                                        \"y\", \"ys\",\n                                                        None::new(SourceInformation::dummy()),\n                                                        None::new(SourceInformation::dummy()),\n                                                        SourceInformation::dummy()))"

---
Ok(
    Case(
        Case {
            type_: Union(
                Union {
                    types: {
                        None(
                            None {
                                source_information: SourceInformation {
                                    source_name: "",
                                    location: Location {
                                        line_number: 0,
                                        column_number: 0,
                                    },
                                    line: "",
                                },
                            },
                        ),
                        Reference(
                            Reference {
                                name: "FirstRest",
                                source_information: SourceInformation {
                                    source_name: "",
                                    location: Location {
                                        line_number: 0,
                                        column_number: 0,
                                    },
                                    line: "",
                                },
                            },
                        ),
                    },
                    source_information: SourceInformation {
                        source_name: "",
                        location: Location {
                            line_number: 0,
                            column_number: 0,
                        },
                        line: "",
                    },
                },
            ),
            name: "$lc$firstRest0",
            argument: Application(
                Application {
                    type_: Function(
                        Function {
                            argument: Reference(
                                Reference {
                                    name: "GenericList",
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                            result: Union(
                                Union {
                                    types: {
                                        None(
                                            None {
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                        Reference(
                                            Reference {
                                                name: "FirstRest",
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                    },
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    function: Variable(
                        Variable {
                            name: "deconstruct",
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    argument: Variable(
                        Variable {
                            name: "xs",
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    source_information: SourceInformation {
                        source_name: "",
                        location: Location {
                            line_number: 0,
                            column_number: 0,
                        },
                        line: "",
                    },
                },
            ),
            alternatives: [
                Alternative {
                    type_: None(
                        None {
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    expression: None(
                        None {
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                },
                Alternative {
                    type_: Reference(
                        Reference {
                            name: "FirstRest",
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                    expression: Let(
                        Let {
                            definitions: [
                                VariableDefinition(
                                    VariableDefinition {
                                        name: "y",
                                        body: Case(
                                            Case {
                                                type_: Any(
                                                    Any {
                                                        source_information: SourceInformation {
                                                            source_name: "",
                                                            location: Location {
                                                                line_number: 0,
                                                                column_number: 0,
                                                            },
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                                name: "$lc$firstRest0",
                                                argument: Application(
                                                    Application {
                                                        type_: Function(
                                                            Function {
                                                                argument: Reference(
                                                                    Reference {
                                                                        name: "FirstRest",
                                                                        source_information: SourceInformation {
                                                                            source_name: "",
                                                                            location: Location {
                                                                                line_number: 0,
                                                                                column_number: 0,
                                                                            },
                                                                            line: "",
                                                                        },
                                                                    },
                                                                ),
                                                                result: Any(
                                                                    Any {
                                                                        source_information: SourceInformation {
                                                                            source_name: "",
                                                                            location: Location {
                                                                                line_number: 0,
                                                                                column_number: 0,
                                                                            },
                                                                            line: "",
                                                                        },
                                                                    },
                                                                ),
                                                                source_information: SourceInformation {
                                                                    source_name: "",
                                                                    location: Location {
                                                                        line_number: 0,
                                                                        column_number: 0,
                                                                    },
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        function: Variable(
                                                            Variable {
                                                                name: "first",
                                                                source_information: SourceInformation {
                                                                    source_name: "",
                                                                    location: Location {
                                                                        line_number: 0,
                                                                        column_number: 0,
                                                                    },
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        argument: Variable(
                                                            Variable {
                                                                name: "$lc$firstRest0",
                                                                source_information: SourceInformation {
                                                                    source_name: "",
                                                                    location: Location {
                                                                        line_number: 0,
                                                                        column_number: 0,
                                                                    },
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        source_information: SourceInformation {
                                                            source_name: "",
                                                            location: Location {
                                                                line_number: 0,
                                                                column_number: 0,
                                                            },
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                                alternatives: [
                                                    Alternative {
                                                        type_: Number(
                                                            Number {
                                                                source_information: SourceInformation {
                                                                    source_name: "",
                                                                    location: Location {
                                                                        line_number: 0,
                                                                        column_number: 0,
                                                                    },
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        expression: Variable(
                                                            Variable {
                                                                name: "$lc$firstRest0",
                                                                source_information: SourceInformation {
                                                                    source_name: "",
                                                                    location: Location {
                                                                        line_number: 0,
                                                                        column_number: 0,
                                                                    },
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    },
                                                ],
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                        type_: Number(
                                            Number {
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                        source_information: SourceInformation {
                                            source_name: "",
                                            location: Location {
                                                line_number: 0,
                                                column_number: 0,
                                            },
                                            line: "",
                                        },
                                    },
                                ),
                                VariableDefinition(
                                    VariableDefinition {
                                        name: "ys",
                                        body: Application(
                                            Application {
                                                type_: Function(
                                                    Function {
                                                        argument: Reference(
                                                            Reference {
                                                                name: "FirstRest",
                                                                source_information: SourceInformation {
                                                                    source_name: "",
                                                                    location: Location {
                                                                        line_number: 0,
                                                                        column_number: 0,
                                                                    },
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        result: Reference(
                                                            Reference {
                                                                name: "GenericList",
                                                                source_information: SourceInformation {
                                                                    source_name: "",
                                                                    location: Location {
                                                                        line_number: 0,
                                                                        column_number: 0,
                                                                    },
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        source_information: SourceInformation {
                                                            source_name: "",
                                                            location: Location {
                                                                line_number: 0,
                                                                column_number: 0,
                                                            },
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                                function: Variable(
                                                    Variable {
                                                        name: "rest",
                                                        source_information: SourceInformation {
                                                            source_name: "",
                                                            location: Location {
                                                                line_number: 0,
                                                                column_number: 0,
                                                            },
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                                argument: Variable(
                                                    Variable {
                                                        name: "$lc$firstRest0",
                                                        source_information: SourceInformation {
                                                            source_name: "",
                                                            location: Location {
                                                                line_number: 0,
                                                                column_number: 0,
                                                            },
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                        type_: List(
                                            List {
                                                element: Number(
                                                    Number {
                                                        source_information: SourceInformation {
                                                            source_name: "",
                                                            location: Location {
                                                                line_number: 0,
                                                                column_number: 0,
                                                            },
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                                source_information: SourceInformation {
                                                    source_name: "",
                                                    location: Location {
                                                        line_number: 0,
                                                        column_number: 0,
                                                    },
                                                    line: "",
                                                },
                                            },
                                        ),
                                        source_information: SourceInformation {
                                            source_name: "",
                                            location: Location {
                                                line_number: 0,
                                                column_number: 0,
                                            },
                                            line: "",
                                        },
                                    },
                                ),
                            ],
                            expression: None(
                                None {
                                    source_information: SourceInformation {
                                        source_name: "",
                                        location: Location {
                                            line_number: 0,
                                            column_number: 0,
                                        },
                                        line: "",
                                    },
                                },
                            ),
                            source_information: SourceInformation {
                                source_name: "",
                                location: Location {
                                    line_number: 0,
                                    column_number: 0,
                                },
                                line: "",
                            },
                        },
                    ),
                },
            ],
            source_information: SourceInformation {
                source_name: "",
                location: Location {
                    line_number: 0,
                    column_number: 0,
                },
                line: "",
            },
        },
    ),
)
